docs/flip/2434-general-metadata.md

```diff
@@ -0,0 +1,143 @@
+
+
+## Proposal
+
+To generalize metadata tracking and keep the specific metadata out of core Flax we propose the following abstract base class:
+
+```python
+class AxisMetadata(metaclass=abc.ABCMeta):
+
+  @abc.abstractmethod
+  def unbox(self) -> Any:
+    pass
+
+  @abc.abstractmethod
+  def add_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
+    pass
+
+  @abc.abstractmethod
+  def remove_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
+    pass
+```
+
+We call this type of class wrapping a value and keeping track of some additional data a **box**.
+By defining an abstract base class for this box, the API does not need to be aware of the specifics of the metadata that is tracked.
+This should make the API future proof and modular.
+
+The ``add_axis`` and ``remove_axis`` callback return an instance of their own type instead of mutating in-place.
+Typically, an implementation would be a ``flax.struct.PyTreeNode`` because the box should still be a valid JAX value and must therefore be handled by the PyTree API.
+Calling ``jax.tree_map`` on a boxed value will simply map over the value in the box.
+The lifted transforms that need to handle metadata will call ``jax.tree_map(..., is_leaf=lambda x: isinstance(x, AxisMetadata))`` to find the AxisMetadata instances within a PyTree.
+
+Advantages of the boxing approach:
+1. Boxing can be used outside of Flax and metadata is automatically "inherited". For example, the optimizerd state will
+   have the same partitioning spec as the parameters, because the state is initialized using a ``jax.tree_map`` over the boxed parameters.
+2. Boxes are composable.
+3. Boxing avoids string manipulation and generally avoids having to handle additional auxilary collections like "param_axes" in the current
+   partitioning API.
+4. No need to lift metadata collections seperately.
+
+
+Disadvantages:
+1. Handling boxed values requires the relatively new ``is_leaf=`` syntax which users might not be familiar with. Although users will
+   probably call Flax provided utils that handle the low-level tree_map calls in most cases.
+3. Custom Pytree nodes have a small runtime overhead. It's hard to observe this in practise because JAX calls are async.
+
+
+### Init syntax
+
+
+Boxes can be created directly by the init function of a variable. Therefore, we propose to create metadata using higher-order initializers.
+The main advantage of this is that we can decouple metadata handling compltely from the Module definition. Also, most Modules already overwrite
+attributes to override the default initialzers so users can add metadata to existing Modules without requiring any code changes.
+
+To illustrate this, let's consider a metadata class that keeps track of PartitionSpecs used by ``pjit``:
+
+```python
+class Partitioned(flax.struct.PyTreeNode, AxisMetadata):
+  value: Any
+  named_axes: Tuple[Optional[str], ...]
```

Comment:
wouldn't this metadata usually have `flax.struct.field(pytree_node=False)` ?

Reply:
Yes

Reply:
correct